File: archive/pulumi-py/Pr6231/lib/alb.py
"""
Application Load Balancer
Creates ALB with target group and listener rules
"""

import pulumi
import pulumi_aws as aws

def create_alb(environment_suffix: str, vpc, public_subnets, security_group):
    """
    Create Application Load Balancer with target group
    """

    # Create ALB
    alb = aws.lb.LoadBalancer(
        f"flask-alb-{environment_suffix}",
        name=f"flask-alb-{environment_suffix}",
        load_balancer_type="application",
        subnets=[subnet.id for subnet in public_subnets],
        security_groups=[security_group.id],
        enable_deletion_protection=False,
        enable_http2=True,
        enable_cross_zone_load_balancing=True,
        tags={
            "Name": f"flask-alb-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Create target group
    target_group = aws.lb.TargetGroup(
        f"flask-tg-{environment_suffix}",
        name=f"flask-tg-{environment_suffix}",
        port=5000,
        protocol="HTTP",
        vpc_id=vpc.id,
        target_type="ip",
        deregistration_delay=30,
        health_check=aws.lb.TargetGroupHealthCheckArgs(
            enabled=True,
            healthy_threshold=2,
            unhealthy_threshold=3,
            timeout=5,
            interval=30,
            path="/health",
            protocol="HTTP",
            matcher="200"
        ),
        tags={
            "Name": f"flask-tg-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Create default listener
    listener = aws.lb.Listener(
        f"flask-listener-{environment_suffix}",
        load_balancer_arn=alb.arn,
        port=80,
        protocol="HTTP",
        default_actions=[
            aws.lb.ListenerDefaultActionArgs(
                type="forward",
                target_group_arn=target_group.arn
            )
        ]
    )

    # Create listener rule for /api/* path
    api_rule = aws.lb.ListenerRule(
        f"flask-api-rule-{environment_suffix}",
        listener_arn=listener.arn,
        priority=100,
        conditions=[
            aws.lb.ListenerRuleConditionArgs(
                path_pattern=aws.lb.ListenerRuleConditionPathPatternArgs(
                    values=["/api/*"]
                )
            )
        ],
        actions=[
            aws.lb.ListenerRuleActionArgs(
                type="forward",
                target_group_arn=target_group.arn
            )
        ]
    )

    # Create listener rule for /health path
    health_rule = aws.lb.ListenerRule(
        f"flask-health-rule-{environment_suffix}",
        listener_arn=listener.arn,
        priority=101,
        conditions=[
            aws.lb.ListenerRuleConditionArgs(
                path_pattern=aws.lb.ListenerRuleConditionPathPatternArgs(
                    values=["/health"]
                )
            )
        ],
        actions=[
            aws.lb.ListenerRuleActionArgs(
                type="forward",
                target_group_arn=target_group.arn
            )
        ]
    )

    return {
        "alb": alb,
        "target_group": target_group,
        "listener": listener
    }

File: archive/pulumi-ts/Pr7161/lib/components/alb.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface AlbComponentArgs {
  environment: string;
  vpcId: pulumi.Output<string>;
  publicSubnetIds: pulumi.Output<string>[];
  tags: { [key: string]: string };
}

export class AlbComponent extends pulumi.ComponentResource {
  public readonly alb: aws.lb.LoadBalancer;
  public readonly targetGroup: aws.lb.TargetGroup;
  public readonly listener: aws.lb.Listener;
  public readonly dnsName: pulumi.Output<string>;
  public readonly targetGroupArn: pulumi.Output<string>;
  public readonly securityGroupId: pulumi.Output<string>;
  private readonly securityGroup: aws.ec2.SecurityGroup;

  constructor(
    name: string,
    args: AlbComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:payment:AlbComponent', name, {}, opts);

    const resourceOpts = { parent: this };

    // Create security group for ALB
    this.securityGroup = new aws.ec2.SecurityGroup(
      `${args.environment}-payment-alb-sg`,
      {
        name: `${args.environment}-payment-alb-sg`,
        description: 'Security group for Application Load Balancer',
        vpcId: args.vpcId,
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'HTTP access',
          },
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'HTTPS access',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...args.tags,
          Name: `${args.environment}-payment-alb-sg`,
        },
      },
      resourceOpts
    );

    this.securityGroupId = this.securityGroup.id;

    // Create Application Load Balancer
    this.alb = new aws.lb.LoadBalancer(
      `${args.environment}-payment-alb`,
      {
        name: `${args.environment}-payment-alb`,
        internal: false,
        loadBalancerType: 'application',
        securityGroups: [this.securityGroup.id],
        subnets: args.publicSubnetIds,
        enableDeletionProtection: false,
        enableHttp2: true,
        enableCrossZoneLoadBalancing: true,
        tags: {
          ...args.tags,
          Name: `${args.environment}-payment-alb`,
        },
      },
      resourceOpts
    );

    this.dnsName = this.alb.dnsName;

    // Create target group
    this.targetGroup = new aws.lb.TargetGroup(
      `${args.environment}-payment-tg`,
      {
        name: `${args.environment}-payment-tg`,
        port: 8080,
        protocol: 'HTTP',
        vpcId: args.vpcId,
        targetType: 'ip',
        healthCheck: {
          enabled: true,
          path: '/health',
          protocol: 'HTTP',
          matcher: '200',
          interval: 30,
          timeout: 5,
          healthyThreshold: 2,
          unhealthyThreshold: 3,
        },
        deregistrationDelay: 30,
        tags: {
          ...args.tags,
          Name: `${args.environment}-payment-tg`,
        },
      },
      resourceOpts
    );

    this.targetGroupArn = this.targetGroup.arn;

    // Create listener with path-based routing
    this.listener = new aws.lb.Listener(
      `${args.environment}-payment-listener`,
      {
        loadBalancerArn: this.alb.arn,
        port: 80,
        protocol: 'HTTP',
        defaultActions: [
          {
            type: 'forward',
            targetGroupArn: this.targetGroup.arn,
          },
        ],
        tags: args.tags,
      },
      resourceOpts
    );

    // Add path-based routing rules
    new aws.lb.ListenerRule(
      `${args.environment}-payment-api-rule`,
      {
        listenerArn: this.listener.arn,
        priority: 100,
        actions: [
          {
            type: 'forward',
            targetGroupArn: this.targetGroup.arn,
          },
        ],
        conditions: [
          {
            pathPattern: {
              values: ['/api/*'],
            },
          },
        ],
        tags: args.tags,
      },
      resourceOpts
    );

    new aws.lb.ListenerRule(
      `${args.environment}-payment-webhook-rule`,
      {
        listenerArn: this.listener.arn,
        priority: 200,
        actions: [
          {
            type: 'forward',
            targetGroupArn: this.targetGroup.arn,
          },
        ],
        conditions: [
          {
            pathPattern: {
              values: ['/webhook/*'],
            },
          },
        ],
        tags: args.tags,
      },
      resourceOpts
    );

    this.registerOutputs({
      dnsName: this.dnsName,
      targetGroupArn: this.targetGroupArn,
      securityGroupId: this.securityGroupId,
    });
  }

  // Helper method to get security group ID
  public getSecurityGroupId(): pulumi.Output<string> {
    return this.securityGroupId;
  }
}
